source/MRMesh/MR2DContoursTriangulation.cpp

```diff
@@ -61,217 +212,637 @@ class PlanarTriangulator
+            mesh_.topology.splice( lowerLone, newEdge );
+            mesh_.topology.splice( connectorEdgeId, newEdge.sym() );
 
-        if ( ccwBI && !ccwBIn )
-            return left ? i : nextI;
-        if ( !ccwBI && ccwBIn )
-            return left ? nextI : i;
+            lowerLone = upperLone = {};
 
-        pvc[0] = coordI;
-        bool ccwIIn = ccw( pvc );
+            windingInfo_.autoResizeSet( newEdge.undirected(), windingInfo_[activeSweepEdges_[i].edgeId.undirected()] );
+            if ( i == minIndex - 1 )
+                lowestLeft = newEdge;
+        }
+    }
+    if ( numRight == 0 )
+    {
+        if ( stage_ == Stage::Monotonation && minIndex > 0 && maxIndex + 1 < activeSweepEdges_.size() &&
+            windingInfo_[activeSweepEdges_[minIndex - 1].edgeId.undirected()].inside( windingMode_ ) )
+        {
+            activeSweepEdges_[minIndex - 1].upperInfo.loneEdgeId = lowestLeft.sym();
+            activeSweepEdges_[maxIndex + 1].lowerInfo.loneEdgeId = lowestLeft.sym();
+        }
+        activeSweepEdges_.erase( activeSweepEdges_.begin() + minIndex, activeSweepEdges_.begin() + maxIndex + 1 );
+        if ( stage_ == Stage::Intersections )
+        {
+            checkIntersection_( minIndex - 1, false );
+        }
+    }
+    else
+    {
+        for ( int i = minIndex; i < minIndex + std::min( numLeft, numRight ); ++i )
+        {
+            assert( i < activeSweepEdges_.size() );
+            activeSweepEdges_[i] = rightGoingCache_[i - minIndex];
+        }
+        if ( numLeft > numRight )
+            activeSweepEdges_.erase( activeSweepEdges_.begin() + minIndex + numRight, activeSweepEdges_.begin() + maxIndex + 1 );
+        else if ( numLeft < numRight )
+            activeSweepEdges_.insert( activeSweepEdges_.begin() + maxIndex + 1, rightGoingCache_.begin() + numLeft, rightGoingCache_.end() );
 
-        if ( ccwBI && ccwIIn )
-            return left ? i : nextI;
-        if ( !ccwBI && !ccwIIn )
-            return left ? nextI : i;
+        if ( stage_ == Stage::Intersections )
+        {
+            checkIntersection_( minIndex + numRight, true );
+            checkIntersection_( minIndex - 1, false );
+        }
+    }
+}
 
-        if ( nextI == 1 )
-            break;
-        i = nextI;
+void SweepLineQueue::processIntersectionEvent_( int index )
+{
+    bool isValid = isIntersectionValid_( index );
+    if ( isValid )
+    {
+        intersections_.emplace_back( Intersection{
+            .lower = activeSweepEdges_[index].edgeId,
+            .upper = activeSweepEdges_[index + 1].edgeId } );
     }
-    assert( false );
-    return 0;
+    invalidateIntersection_( index );
+    if ( !isValid )
+        return;
+
+    auto minEdgeId = std::min( activeSweepEdges_[index].edgeId, activeSweepEdges_[index + 1].edgeId );
+    auto maxEdgeId = std::max( activeSweepEdges_[index].edgeId, activeSweepEdges_[index + 1].edgeId );
+
+    auto& interInfo = intersectionsMap_.at( { minEdgeId,maxEdgeId } );
```

Comment:
`at()` method can throw exceptions. Either catch them or use not-throwing method